### PR TITLE
syncer: add rolling update feature code annotation #165

### LINKED ISSUE #165: [feature] add rolling update feature code annotation
**Is your feature request related to a problem? Please describe.**

Detailed description of the implementation process of the rolling update feature through code annotation。
for pull : *: support roll update #133

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by runkecheng**
### What type of PR is this?

/enhancement

### Which issue(s) this PR fixes?

Fixes #165

### What this PR does?

Summary: add rolling update feature code annotation

### CODE REVIEW

File: cluster/syncer/statefulset.go [hunk 1]
```diff
@@ -105,14 +106,19 @@ func (s *StatefulSetSyncer) Sync(ctx context.Context) (syncer.SyncResult, error)
 
 	result.Operation, err = s.createOrUpdate(ctx)
 
+	//get namespace.
```
Comment (acekingke) [inline cluster/syncer/statefulset.go:109]: key is `ObjectKey{Namespace: obj.GetNamespace(), Name: obj.GetName()}` , it contain namespace and name

Comment (runkecheng) [reply cluster/syncer/statefulset.go]: fixed.

**Review by zhyass — CHANGES_REQUESTED** [review]
The first letter should be capitalized.

File: cluster/syncer/statefulset.go [hunk 2]
```diff
@@ -199,31 +207,33 @@ func (s *StatefulSetSyncer) updatePod(ctx context.Context) error {
...
-
+	// All followers have been updated now, then update leader.
 	if leaderPod.Name != "" {
+		// When replicas is two (one leader and one follower).
 		if len(followerPods) == 1 {
 			if err := s.preUpdate(ctx, leaderPod.Name, followerPods[0].Name); err != nil {
 				return err
 			}
 		}
+		// Replicas greater than two.
```
Comment (zhyass) [inline cluster/syncer/statefulset.go:236]: Update the leader. Replicas=2 also need run the code.

File: cluster/syncer/statefulset.go [hunk 3]
```diff
@@ -232,7 +242,10 @@ func (s *StatefulSetSyncer) updatePod(ctx context.Context) error {
 	return nil
 }
 
-// preUpdate run before update the leader pod when replica is 2.
+// preUpdate run before update the leader pod when replicas is 2.
+// For a two-node cluster, we need to adjust the parameter rpl_semi_sync_master_timeout
```
Comment (zhyass) [inline cluster/syncer/statefulset.go:246]: Wrong describe. `rpl_semi_sync_master_timeout` is automatically set by xenon.
`preUpdate` used to manually switch the master node.

File: cluster/syncer/statefulset.go [hunk 4]
```diff
@@ -133,9 +139,10 @@ func (s *StatefulSetSyncer) Sync(ctx context.Context) (syncer.SyncResult, error)
 }
 
 // createOrUpdate creates or updates the statefulset in the Kubernetes cluster.
-// see https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.2/pkg/controller/controllerutil?utm_source=gopls#CreateOrUpdate
+// See https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.2/pkg/controller/controllerutil?utm_source=gopls#CreateOrUpdate
 func (s *StatefulSetSyncer) createOrUpdate(ctx context.Context) (controllerutil.OperationResult, error) {
 	var err error
+	// Check if sfs exist.
```
Comment (zhyass) [inline cluster/syncer/statefulset.go:145]: `Check if statefulset exists` or `Check if s.sfs exists`

File: cluster/syncer/statefulset.go [hunk 5]
```diff
@@ -480,9 +492,7 @@ func (s *StatefulSetSyncer) applyNWait(ctx context.Context, pod *corev1.Pod) err
 	})
 }
 
-// retry runs func "f" every "in" time until "limit" is reached
-// it also doesn't have an extra tail wait after the limit is reached
-// and f func runs first time instantly
+// retry runs func "f" every "in" time until "limit" is reached.
```
Comment (zhyass) [inline cluster/syncer/statefulset.go:495]: Why remove the annotation.

Comment (runkecheng) [reply cluster/syncer/statefulset.go:501]: sorry, I add it back

**Review by zhyass — APPROVED** [review]

**Review by zhyass — COMMENTED** [review]
LGTM

File: cluster/syncer/statefulset.go [hunk 6]
```diff
@@ -105,14 +106,19 @@ func (s *StatefulSetSyncer) Sync(ctx context.Context) (syncer.SyncResult, error)
...
 	gvk, gvkErr := apiutil.GVKForObject(s.sfs, s.cli.Scheme())
 	if gvkErr != nil {
 		kind = fmt.Sprintf("%T", s.sfs)
 	} else {
 		kind = gvk.String()
 	}
-
+	// Print log.
+	// Info: owner is deleted or ignored error.
+	// Waring: other errors.
```
Comment (andyli029) [inline cluster/syncer/statefulset.go:120]: Waring
->
Warning

**Pull request merged**